src/ir/typescript_types.py

```diff
@@ -0,0 +1,201 @@
+    def get_builtin(self):
+        return TypeScriptBuiltin
+
+    def get_void_type(self):
+        return VoidType()
+
+    def get_any_type(self):
+        return ObjectType()
+
+    def get_number_type(self):
+        return NumberType(primitive=False)
+    
+    def get_boolean_type(self):
+        return BooleanType(primitive=False)
+
+    def get_char_type(self):
+        return StringType(primitive=False)
+
+    def get_string_type(self):
+        return StringType(primitive=False)
+    
+    def get_big_integer_type(self):
+        return BigIntegerType(primitive=False)
+
+    def get_array_type(self):
+        return ArrayType()
+
+    def get_function_type(self, nr_parameters=0):
+        return FunctionType(nr_parameters)
+    
+    def get_object_type(self):
+        return ObjectLowercaseType()
+    
+    def get_primitive_types(self):
+        return [
+            NumberType(primitive=True),
+            StringType(primitive=True),
+            SymbolType(primitive=True),
+            BooleanType(primitive=True),
+            BigIntegerType(primitive=True)
+        ]
+
+    def get_integer_type(self):
+        return NumberType(primitive=False)
+
+    def get_byte_type(self):
+        return NumberType(primitive=False)
+
+    def get_short_type(self):
+        return NumberType(primitive=False)
+
+    def get_long_type(self):
+        return NumberType(primitive=False)
+
+    def get_float_type(self):
+        return NumberType(primitive=False)
+
+    def get_double_type(self):
+        return NumberType(primitive=False)
+
+    def get_big_decimal_type(self):
+        return NumberType(primitive=False)
+
+
+class TypeScriptBuiltin(Builtin):
+    def __init__(self, name, primitive):
+        super().__init__(name)
+        self.primitive = primitive
+    
+    def __str__(self):
+        if not self.is_primitive():
+            return str(self.name) + "(typescript-builtin)"
+        return str(self.name).lower() + "(typescript-primitive)"
+    
+    def is_primitive(self):
+        return self.primitive
+
+
+class ObjectType(TypeScriptBuiltin):
+    def __init__(self, name="Object"):
+        super().__init__(name, False)
+
+class ObjectLowercaseType(TypeScriptBuiltin):
+    def __init__(self, name="object"):
+        super().__init__(name, False)
+        self.supertypes.append(ObjectType())
+
+
+class VoidType(TypeScriptBuiltin):
+    def __init__(self, name="void"):
+        super().__init__(name, False)
+        self.supertypes.append(ObjectType())
+
+
+class NumberType(TypeScriptBuiltin):
+    def __init__(self, name="Number", primitive=False):
+        super().__init__(name, primitive)
+        self.supertypes.append(ObjectType())
+    
+    def is_assignable(self, other):
+        return type(other) is NumberType
+
+    def box_type(self):
+        return NumberType(self.name, primitive=False)
+    
+    def get_name(self):
+        if self.is_primitive:
+            return "number"
+        return super().get_name()
+
```

Comment:
add a newline here. There should be 2 newlines among top-level definitions, and one newline among class methods.